java/libs/src/main/java/org/physical_web/collection/PhysicalWebCollection.java

```diff
@@ -260,6 +271,38 @@ public void jsonDeserialize(JSONObject jsonObject) throws PhysicalWebCollectionE
   }
 
   /**
+   * Return a list of UrlGroups sorted by the rank of the top-ranked pair in each group.
+   * @return a sorted list of UrlGroups.
+   */
+  public List<UrlGroup> getUrlGroupsSortedByRank() {
+    List<PwPair> sortedPwPairs = getPwPairsSortedByRank();
+
+    Map<String, UrlGroup> pwGroupMap = new HashMap<>();
+    for (PwPair pwPair : sortedPwPairs) {
+      PwsResult pwsResult = pwPair.getPwsResult();
+      if (pwsResult != null) {
+        String groupId = pwsResult.getGroupId();
+        if (groupId != null && !groupId.equals("")) {
```

Comment:
I think we can assume empty is valid (though current PWS will always return a result) and still get the api that we want.

Just create an empty group for each PwPair that doesn't have a groupid.

I think this api is meant to be: get me all valid results, but group them.